src/messaging/ExchangeHolder.h

```diff
@@ -0,0 +1,136 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/Optional.h>
+#include <lib/support/IntrusiveList.h>
+#include <messaging/ExchangeContext.h>
+
+namespace chip {
+namespace Messaging {
+
+/** @brief
+ *      This provides a wrapper around ExchangeContext that automatically manages
+ *      cleaning up the EC without any extra involvement. This is meant to be used
+ *      by application and protocol logic code that would otherwise need to closely
+ *      manange their internal pointers to an ExchangeContext and correctly
+ *      null-it out/abort it depending on the circumstances. This relies on clear rules
+ *      established by ExchangeContext and the transfer of ownership at various points
+ *      in its lifetime.
+ *
+ *      It does this by listening in on OnExchangeClosing and looking at the various
+ *      states the exchange might be in to decide amongst a couple of different tactics.
+ *
+ *      This is a delegate forwarder - consumers can still register to be an ExchangeDelegate
+ *      and get notified of all relevant happenings on that delegate interface.
+ *
+ */
+class ExchangeHolder : public ExchangeDelegate
+{
+public:
+    ExchangeHolder(ExchangeDelegate & delegate) : mpExchangeDelegate(delegate) {}
+
+    virtual ~ExchangeHolder() { AbortIfNeeded(); }
+
+    bool Contains(const ExchangeContext * exchange) const { return mpExchangeCtx == exchange; }
+
+    /*
+     *  Aborts any previously tracked exchange and
+     *  re-acquires the provided exchange. As part of that,
+     *  this will set itself as the delegate on that exchange (and forward
+     *  calls to the registered delegate on this object).
```

Comment:
When would you do that?

Reply:
nit: please make this doxygen comment format: `/** ... */`

I'm okay with the name `Grab`.  Agree that the comment could be improved with regard to usage instructions:

"Replaces the held Exchange and associated delegate to instead track the given ExchangeContext, aborting and dereferencing any previously held Exchange as necessary.  This method is called by the upper layer whenever the protocol logic is transitioning from an outdated Exchange to a new one, often during the start of a new transaction."

Reply:
Will take Martin's suggestion.

Reply:
This class is modeled after a similar, weak-pointer like construct in sessions called `SessionHolder`. That has a similar `Grab` method as well. Mostly, I'm trying to keep consistency in the SDK here.